[WIP] ASB transport adapter sample

SzymonPobiega opened the pull request:
- [x] Make the sample more real-life by adding some complexity (e.g. body encoding + secure connection strings)
 - [x] Updated `sample.md`
- [x] Review

SzymonPobiega commented:
So far I managed to implement a simple case where both endpoints use a single namespace while the SC uses a different one. The endpoints use the new forwarding topology while the SC uses the old one (but it should not matter since we don't really do pubsub there).

@SeanFeldman I've looked at the doco and asked on #azure and if I understand correctly the cross-namespace routing (https://docs.particular.net/transports/azure-service-bus/multiple-namespaces-support#cross-namespace-routing) does not work for pub/sub. So a scenario where each endpoint uses a different namespace seems weird to me (since the endpoints could not do pub/sub). 

Based on that I think it makes sense to make the sample more real-life either by using multi-namespace with round robin or some other feature like the body encoding.

SeanFeldman commented:
We'll do body encoding + secure connection strings.

SeanFeldman commented:
**Update**

- I've updated the sample to use Secure Connection Strings and Body Encoding `Stream` to demonstrate how the business endpoints can follow the recommendation we're giving for the transport while continue using SC.
- Updated sample.md with the changes for the transport and relevant snippets.

@SzymonPobiega please review

SzymonPobiega commented:
@SeanFeldman I reviewed the code once again and figured that we need to go multi-namespace. I did it in 0d071f1a8d1ddbe50c7b1bae6701f7d97795e66c. The reason is, without that the `UseNamespaceAliasesInsteadOfConnectionStrings` is not really used and messages flow correctly. When I switched `Shipping` to use a separate namespace it all fell apart.
 * The destination address for the retry did not include the namespace alias because `FailedQ` header does not include it
 * The `ReplyToAddress` header gets corrupted by ASB 6.x in ServiceControl and the namespace alias is replaced by SC's connection string.

In https://github.com/Particular/ServiceControl.TransportAdapter/pull/10 I added two features required to make it work: destination override based on a custom header included in the failed message and header value restoration.

For the context, here's a link to my slack conversation with Yves https://particularsoftware.slack.com/archives/C1607JM6G/p1500490738006461

SimonCropp commented on samples/servicecontrol/adapter-asb-multi-namespace/SCTransportAdapter_1/Adapter/Adapter.csproj:
```diff
@@ -0,0 +1,64 @@
+    <Reference Include="System.Data" />
+    <Reference Include="System.Runtime.Serialization" />
+    <Reference Include="System.ServiceModel" />
+    <Reference Include="System.Xml" />
```
is the xml ref required?

SimonCropp commented on samples/servicecontrol/adapter-asb-multi-namespace/diagram.svg:
```diff
@@ -0,0 +1 @@
+<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="1160" height="921.33"><g transform="translate(-60 -60)"><path d="M80 80h160v160H80z" fill="#ef8d22"/><path d="M181.87 116.8v18.13H160v25.6h21.87v18.14H120V116.8z" fill="#fff"/><path d="M163.73 138.67H200v18.1h-36.27z" fill="#fff"/><path d="M80 205c0-2.76 2.24-5 5-5h150c2.76 0 5 2.24 5 5v30c0 2.76-2.24 5-5 5H ...
```
can we use mermaid instead?

SeanFeldman replied:
I'll give it a try @SimonCropp

SeanFeldman commented:
I'm back and will review this tomorrow.

SeanFeldman commented:
@SzymonPobiega I've renamed adapter specific headers to `"_adapter.Original...."` to indicate better that those are *not* NServiceBus specific header.
![image](https://user-images.githubusercontent.com/1309622/28596830-0964ca10-7158-11e7-85fa-fde0532ff635.png)

Also, in order to visualize each endpoints` failure simulation status, added a bit of sugar icing
![image](https://user-images.githubusercontent.com/1309622/28597151-991603f8-7159-11e7-9b1a-8d2f53b54833.png)

SeanFeldman commented:
This is what the diagram looks like now:

![image](https://user-images.githubusercontent.com/1309622/28598096-aca10666-715e-11e7-952d-bf2b66ee5a97.png)

Since I've change things, @SimonCropp could you please do a last review? Thanks.
